mergify_engine/worker.py

```diff
@@ -0,0 +1,366 @@
+            )
+
+            try:
+                await self._run_engine(
+                    installation_id, owner, repo, pull_number, sources
+                )
+            except MaxPullRetry as e:
+                logger.error(
+                    "fail to process pull request, abandoning",
+                    attempts=e.attempts,
+                    exc_info=True,
+                )
+            except MaxStreamRetry as e:
+                logger.error(
+                    "fail to process stream, abandoning",
+                    attempts=e.attempts,
+                    exc_info=True,
+                )
+                break
+            except PullRetry as e:
+                logger.info(
+                    "fail to process pull request, retrying",
+                    attempts=e.attempts,
+                    exc_info=True,
+                )
+                del message_ids[(owner, repo, pull_number)]
+            except StreamRetry as e:
+                logger.info(
+                    "fail to process stream, retrying",
+                    attempts=e.attempts,
+                    retry_at=e.retry_at,
+                    exc_info=True,
+                )
+                return
+            except Exception:
+                logger.error("fail to process pull request", exc_info=True)
+
+        message_ids_to_delete = list(
+            itertools.chain.from_iterable(message_ids.values())
+        )
+        score = utils.utcnow().timestamp()
+        await self._redis.eval(
+            self.ATOMIC_CLEAN_STREAM_SCRIPT,
+            keys=[stream_name.encode()],
+            args=[score, len(message_ids_to_delete)] + message_ids_to_delete,
+        )
+
+    # NOTE(sileht): If the stream still have messages, we update the score to reschedule the
+    # pull later
+    ATOMIC_CLEAN_STREAM_SCRIPT = """
+local stream_name = KEYS[1]
+local score = ARGV[1]
+local message_ids_len =  tonumber(ARGV[2])
+local message_ids = {unpack(ARGV, 3, message_ids_len + 3)}
+
+if message_ids_len > 0 then
+    redis.call("XDEL", stream_name, unpack(message_ids))
+end
+
+local len = tonumber(redis.call('XLEN', stream_name))
+if len == 0 then
+    redis.call("ZREM", "streams", stream_name)
+    redis.call("DEL", stream_name)
+else
+    redis.call("ZADD", "streams", score, stream_name)
+end
+"""
+
+
+class Worker:
+    def __init__(self, worker_count=config.STREAM_WORKERS, idle_sleep_time=0.42):
+        self._idle_sleep_time = idle_sleep_time
+        self._worker_count = worker_count
+        self._worker_tasks = []
+        self._stream_processor = None
+
+        self._loop = asyncio.get_running_loop()
+        self._running = asyncio.Event()
+
+    async def stream_worker_task(self, worker_id):
+        # NOTE(sileht): This task must never fail, we don't want to write code to
+        # reap/clean/respawn them
+        while self._running.is_set():
+            try:
+
+                async with self._stream_processor.next() as stream_name:
```

Comment:
It's a bit weird to use `next` as a name as it's usually used for iterators.

Reply:
Yeah, if you have a better one, I take it.